Userland/Libraries/LibGfx/AntiAliasingPainter.cpp

```diff
@@ -169,3 +170,218 @@ void Gfx::AntiAliasingPainter::draw_cubic_bezier_curve(const FloatPoint& control
         draw_line(fp1, fp2, color, thickness, style);
     });
 }
+
+void Gfx::AntiAliasingPainter::draw_circle(IntPoint center, int radius, Color color)
+{
+    /*
+    Algorithm from: https://cs.uwaterloo.ca/research/tr/1984/CS-84-38.pdf
+    Inline comments are from the paper.
+    */
+
+    // TODO: Generalize to ellipses (see paper)
+
+    constexpr int subpixel_resoulution = 255;
+    const int intensity { color.alpha() };
+
+    // Note: Variable names below are from the paper
+
+    /* Current pixel address */
+    int i { 0 }, q { radius };
+
+    /* 1st and 2nd order differences of y*/
+    int delta_y { 0 }, delta2_y { 0 };
+
+    /* Exact and predicted values of f(i) */
+    int y { subpixel_resoulution * radius }, y_hat { 0 };
+
+    /* The value of f(i)*f(i) */
+    int f2 { y * y };
+
+    /* 1st and 2nd order differences of f(i)*f(i) */
+    int delta_f2 { subpixel_resoulution * subpixel_resoulution },
+        delta2_f2 { -delta_f2 - delta_f2 };
+
+    /* v/subpixel_resoulution = percentage of pixel intersected by circle */
+    int v { 0 }, v_old { 0 };
+
+    auto predict = [&] {
+        delta_y += delta2_y;
+        y_hat = y + delta_y; /* y_hat is the predicted value of f(i) */
+    };
+
+    auto minimize = [&] {
+        /* Initialize the minimization */
+        delta_f2 += delta2_f2;
+        f2 += delta_f2;
+
+        int e = y_hat * y_hat - f2;
+        int d = 1;
+        y = y_hat;
+
+        /* Force e negative */
+        if (e > 0) {
+            e = -e;
+            d = -1;
+        }
+
+        /* Minimize */
+        int e_old;
+        while (e < 0) {
+            y = y + d;
+            e_old = e;
+            e += y + y - d;
+        }
+
+        /* e or e_old is minimum squared error */
```

Comment:
this is not the right place for this comment (and in fact, just rename `e` -> `min_squared_error` or just `error`.